File: sourceFile/Java/101_Symmetric_Tree.java
/*
 * @Author: AlanGolphi
 * @Date: 2020-09-06 23:20:08
 * @LastEditTime: 2020-09-06 23:32:49
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        return check(root, root);
    }
    public boolean check(TreeNode p1, TreeNode p2) {
        if (p1 == null && p2 == null) return true;
        if (p1 == null || p2 == null) return false;
        return(p1.val == p2.val && check(p1.left, p2.right) && check(p1.right,p2.left));
    }
}

//递归法，判断一棵树的左节点是否等于右节点
//会报不知名的错 ：java.lang.NullPointerException
//仔细检查过了，一样的代码，先挖个坑，以后慢慢看看。

File: sourceFile/typescript/101_SymmetricTree.ts
import { TreeNode, createTreeFromArray } from "./models";

// @algorithm @lc id=101 lang=typescript
// @title symmetric-tree
// @test([1,2,2,3,4,4,3])=true
// @test([1,2,2,null,3,null,3])=false
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isSymmetric(root: TreeNode | null): boolean {
  if (!root) return true;

  const compare = (p: TreeNode | null, q: TreeNode | null): boolean => {
    if (p == null && q == null) return true;
    if (p == null || q == null) return false;

    return (
      p.val === q.val && compare(p.left, q.right) && compare(p.right, q.left)
    );
  };

  return compare(root.left, root.right);
}

const tree1 = createTreeFromArray([1, 2, 2, 3, 4, 4, 3]);
const tree2 = createTreeFromArray([1, 2, 2, null, 3, null, 3]);

console.log("tree1: ", isSymmetric(tree1));
console.log("tree2: ", isSymmetric(tree2));
